peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+
+::
+
+   STRS_TYPE_NAME: Literal['str', 'list[str]'] = 'str'
+   STRS_TYPE: TypeExpr = STRS_TYPE_NAME  # ERROR: Literal[] value is not a TypeExpr
+
+However ``Literal[...]`` itself is still a ``TypeExpr``:
+
+::
+
+   DIRECTION_TYPE: TypeExpr = Literal['left', 'right']  # OK
+
+
+.. _non_universal_typeexpr:
+
+Non-Universal TypeExprs
+'''''''''''''''''''''''
+
+``TypeExpr[]`` values include *all* type expressions including some
+which are not valid in all annotation contexts. In particular:
+
+-  ``Self`` (valid only in some contexts)
+-  ``TypeGuard[...]`` (valid only in some contexts)
+-  ``TypeIs[...]`` (valid only in some contexts)
+
+``Self`` can be used as a function parameter type, a return type, or a
+variable type, but only within a class definition:
+
+::
+
+   class ReturnsSelf:
+       def __call__(self: Self) -> Self:  # OK
+           return self
+
+   class HoldsSelf:
+       me: Self  # OK
+
+   def returns_self(me: Self) -> Self: ...  # ERROR: Self meaningless outside class
+
+``TypeGuard[]`` and ``TypeIs[]`` can only be used as a return type of a
+function. They cannot be used as a variable type or a parameter type:
+
+::
+
+   def is_positive_int(value: object) -> TypeGuard[int]: ...  # OK
+
+   def nonsense(value: TypeGuard[int]): ...  # ERROR: TypeGuard[] not meaningful here
+
+   exotic_bool: TypeGuard[int]  # ERROR: TypeGuard[] not meaningful here
+
+Non-universal type expressions may be treated as equivalent to ``Any``
+for inference purposes:
+
+::
+
+   typx: TypeExpr = TypeIs[str]
+   assert_type(typx, TypeExpr[Any])  # acceptable
+
+Non-universal type expressions may also be treated as equivalent to
+``Any`` when 
+:ref:`constraining type variables <interactions_with_type_variables>`.
+
+
+Static vs. Runtime Representations of TypeExprs
+'''''''''''''''''''''''''''''''''''''''''''''''
+
+A ``TypeExpr`` value appearing statically in a source file may be normalized
+to a different representation at runtime:
+
+-  For example string-based forward references are normalized at runtime
+   to be ``ForwardRef`` instances: [#forward_ref_normalization]_
+
+   ::
+
+      >>> IntTree = list[typing.Union[int, 'IntTree']]
+      >>> IntTree
+      list[typing.Union[int, ForwardRef('IntTree')]]
+
+-  Special forms themselves often have a runtime representation which is
+   different than how they appear in a source file:
+
+   ::
+
+      >>> list[int]
+      list[int]
+      >>> type(list[int])
+      <class 'types.GenericAlias'>
+      >>> types.GenericAlias(list, int) == list[int]
+      True
+
+The runtime representations of ``TypeExpr``\ s are considered implementation
+details that may change over time and therefore static type checkers are
+not required to recognize them:
+
+::
+
+   INT_TREE: TypeExpr = ForwardRef('IntTree')  # ERROR: Runtime-only form
+   LIST_OF_INT: TypeExpr = types.GenericAlias(list, int)  # ERROR: Runtime-only form
+
+However runtime type checkers must still be prepared to recognize and
```

Comment:
"must" is quite strong language here. It almost sounds like you're mandating behaviour for other runtime type checkers. Maybe "must often"?

Reply:
Well, a runtime type checker is by definition working with the _runtime_ representations of any type expressions that it supports. This is not a new requirement on runtime type checkers; it just isn't a requirement for static type checkers.

Reply:
I think I'll replace "must" with "already must", to be clear that I'm not imposing a _new_ requirement.